compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,871 @@
+    pub struct TypeIdOptions: u32 {
+        const NO_OPTIONS = 0;
+        const GENERALIZE_POINTERS = 1;
+        const RESERVED = 2;
+    }
+}
+
+bitflags! {
+    /// Options for encode_ty.
+    struct EncodeTyOptions: u32 {
+        const NO_OPTIONS = 0;
+        const RESERVED = 1;
+        const GENERALIZE_REPR_C = 2;
+    }
+}
+
+bitflags! {
+    /// Options for transform_ty.
+    struct TransformTyOptions: u32 {
+        const NO_OPTIONS = 0;
+        const GENERALIZE_POINTERS = 1;
+        const GENERALIZE_REPR_C = 2;
+    }
+}
+
+/// Converts a number to a disambiguator (see
+/// https://rust-lang.github.io/rfcs/2603-rust-symbol-name-mangling-v0.html).
+fn to_disambiguator(num: u64) -> String {
+    if let Some(num) = num.checked_sub(1) {
+        format!("s{}_", base_n::encode(num as u128, 62))
+    } else {
+        "s_".to_string()
+    }
+}
+
+/// Converts a number to a sequence number (see
+/// https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangle.seq-id).
+fn to_seq_id(num: usize) -> String {
+    if let Some(num) = num.checked_sub(1) {
+        format!("{}", base_n::encode(num as u128, 36).to_uppercase())
+    } else {
+        "".to_string()
+    }
+}
+
+/// Substitutes a component if found in the substitution dictionary (see
+/// https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling-compression).
+fn compress<'tcx>(
+    dict: &mut FxHashMap<(GenericEntity<'tcx>, TyQ), usize>,
+    key: (GenericEntity<'tcx>, TyQ),
+    comp: &mut String,
+) {
+    match dict.get(&key) {
+        Some(num) => {
+            comp.clear();
+            comp.push_str(&format!("S{}_", to_seq_id(*num)));
+        }
+        None => {
+            dict.insert(key, dict.len());
+        }
+    }
+}
+
+// FIXME(rcvalle): Move to compiler/rustc_middle/src/ty/sty.rs after C types work is done, possibly
+// along with other is_c_type methods.
+/// Returns whether a `ty::Ty` is `c_void`.
+fn is_c_void_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> bool {
+    match ty.kind() {
+        ty::Adt(adt_def, ..) => {
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.item_name(def_id));
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                true
+            } else {
+                false
+            }
+        }
+        _ => false,
+    }
+}
+
+/// Encodes a const using the Itanium C++ ABI with vendor extended type qualifiers and types for
+/// Rust types that are not used at the FFI boundary.
+fn encode_const<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    c: Const<'tcx>,
+    dict: &mut FxHashMap<(GenericEntity<'tcx>, TyQ), usize>,
+    options: EncodeTyOptions,
+) -> String {
+    // u5constI<element-type>E[<const-disambiguator>], where <element-type> is
+    // <const-type>, as vendor extended type
```

Comment:
I'll use https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling.literal for this instead.

Reply:
Done.